File: backend/src/main/kotlin/com/example/backend/controllers/EnRayonController.kt
package com.example.backend.controllers

import com.example.backend.dtos.*
import com.example.backend.models.EnRayon
import com.example.backend.models.Rayon
import com.example.backend.services.EnRayonService
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import java.time.LocalDateTime


@RestController
@RequestMapping("/api/en-rayon")
class EnRayonController(private val enRayonService: EnRayonService) {

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/ajouter")
  fun ajouterProduitsEnRayon(@RequestBody produits: List<ProduitEnRayonDto>): ResponseEntity<List<EnRayon>> {
    val enRayonList = enRayonService.ajouterProduitsEnRayon(produits)
    return ResponseEntity.ok(enRayonList)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PutMapping("/mettre-a-jour")
  fun mettreAJourProduitsEnRayon(@RequestBody produits: List<ProduitEnRayonDto>): ResponseEntity<List<EnRayon>> {
    val enRayonList = enRayonService.mettreAJourProduitsEnRayon(produits)
    return ResponseEntity.ok(enRayonList)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/increment-produit-detail")
  fun decrementerStock(@RequestBody data: ProduitDetailIncrementEnRayonDto): ResponseEntity<Map<String, Any?>> {
    val enRayonList = enRayonService.decrementerStock(data.enRayonId!!.toInt(), data.produitDetailId!!.toInt())
    return ResponseEntity.ok(enRayonList)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/par-nom-produit")
  fun getProduitsEnRayonParNomProduit(@RequestParam nomProduit: String): ResponseEntity<List<EnRayon>> =
    ResponseEntity.ok(enRayonService.getProduitsEnRayonParNomProduit(nomProduit))


  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/par-produit")
  fun getProduitsEnRayonParProduitIdt(@RequestParam produitId: Int): ResponseEntity<List<Map<String, Any?>>> =
    ResponseEntity.ok(enRayonService.getProduitsEnRayonParProduitIdt(produitId))

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/par-produit-all")
  fun getProduitsEnRayonParProduitIdt(
    @RequestParam produitId: Int,
    @RequestParam produitType: String
  ): ResponseEntity<List<Map<String, Any?>>> =
    ResponseEntity.ok(enRayonService.getProduitsEnRayonParProduitIdt(produitId))

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/par-nom-rayon")
  fun getProduitsEnRayonParNomRayon(@RequestParam nomRayon: String): ResponseEntity<List<Rayon>> =
    ResponseEntity.ok(enRayonService.getProduitsEnRayonParNomRayon(nomRayon))

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/par-fournisseur")
  fun getProduitsEnRayonParFournisseur(@RequestParam nomFournisseur: String): ResponseEntity<List<EnRayon>> =
    ResponseEntity.ok(enRayonService.getProduitsEnRayonParFournisseur(nomFournisseur))

  /*  @CrossOrigin(origins = ["http://localhost:4200"])
    @PreAuthorize("isAuthenticated()")
    @GetMapping("/par-uniter")
    fun getProduitsEnRayonParUniter(@RequestParam uniter: String): ResponseEntity<List<EnRayon>> =
      ResponseEntity.ok(enRayonService.getProduitsEnRayonParUniter(uniter))*/

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/par-commande")
  fun getProduitsEnRayonParCommande(@RequestParam commandeId: Long): ResponseEntity<List<EnRayon>> =
    ResponseEntity.ok(enRayonService.getProduitsEnRayonParCommande(commandeId))

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/par-date-livraison")
  fun getProduitsEnRayonParIntervalleDateLivraison(
    @RequestParam startDate: LocalDateTime,
    @RequestParam endDate: LocalDateTime
  ): ResponseEntity<List<EnRayon>> =
    ResponseEntity.ok(enRayonService.getProduitsEnRayonParIntervalleDateLivraison(startDate, endDate))

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/par-date-peremption")
  fun getProduitsEnRayonParIntervalleDatePeremption(
    @RequestParam startDate: LocalDateTime,
    @RequestParam endDate: LocalDateTime
  ): ResponseEntity<List<EnRayon>> =
    ResponseEntity.ok(enRayonService.getProduitsEnRayonParIntervalleDatePeremption(startDate, endDate))

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/par-prix-achat")
  fun getProduitsEnRayonParIntervallePrixAchat(
    @RequestParam minPrix: Double,
    @RequestParam maxPrix: Double
  ): ResponseEntity<List<EnRayon>> =
    ResponseEntity.ok(enRayonService.getProduitsEnRayonParIntervallePrixAchat(minPrix, maxPrix))

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/par-prix-vente")
  fun getProduitsEnRayonParIntervallePrixVente(
    @RequestParam minPrix: Double,
    @RequestParam maxPrix: Double
  ): ResponseEntity<List<EnRayon>> =
    ResponseEntity.ok(enRayonService.getProduitsEnRayonParIntervallePrixVente(minPrix, maxPrix))

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/save")
  fun mettreAJourProduitEnRayon(@RequestBody rayonDto: EnRayonDto?): ResponseEntity<EnRayon> {
    val enRayon = enRayonService.mettreAJourProduitEnRayon(rayonDto!!)
    return ResponseEntity.ok(enRayon)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/pageable")
  fun getProduitsEnRayonPageable(
    @RequestParam(required = false) nomProduit: String?,
    @RequestParam(required = false) bientotPerimee: String?,
    @RequestParam(required = false) joursAvantPeremption: String?,
    @RequestParam(required = false) enStock: String?,
    @RequestParam(defaultValue = "0") page: Int,
    @RequestParam(defaultValue = "10") size: Int,
    @RequestParam(defaultValue = "id") sort: String,
    @RequestParam(defaultValue = "desc") direction: String
  ): ResponseEntity<Page<Map<String, Any?>>> {
    val bientotPerimeeBoolean: Boolean? = when (bientotPerimee?.lowercase()) {
      "true" -> true
      "false" -> false
      else -> null
    }
    val enStockBoolean: Boolean? = when (enStock?.lowercase()) {
      "true" -> true
      "false" -> false
      else -> null
    }
    val joursAvantPeremptionInt: Int = joursAvantPeremption?.toIntOrNull() ?: 0
    val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.fromString(direction), sort))
    val result = enRayonService.getProduitsEnRayonPageable(
      nomProduit,
      bientotPerimeeBoolean,
      joursAvantPeremptionInt,
      enStockBoolean,
      pageable
    )
    return ResponseEntity.ok(result)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/pageable/new")
  fun getProduitsEnRayonPageableNew(
    @RequestParam(required = false) nomProduit: String?,
    @RequestParam(required = false) bientotPerimee: String?,
    @RequestParam(required = false) joursAvantPeremption: String?,
    @RequestParam(required = false) enStock: String?,
    @RequestParam(defaultValue = "0") page: String,
    @RequestParam(defaultValue = "10") size: String,
    @RequestParam(defaultValue = "id") sort: String,
    @RequestParam(defaultValue = "desc") direction: String
  ): ResponseEntity<EnRayonPageableCustomDto> {
    val pageNumber = page.toIntOrNull()?.coerceAtLeast(0) ?: 0
    val pageSize = size.toIntOrNull()?.coerceAtLeast(1) ?: 10
    val bientotPerimeeBoolean: Boolean? = when (bientotPerimee?.lowercase()) {
      "true" -> true
      "false" -> false
      else -> null
    }
    val enStockBoolean: Boolean? = when (enStock?.lowercase()) {
      "true" -> true
      "false" -> false
      else -> null
    }
    val joursAvantPeremptionInt: Int = joursAvantPeremption?.toIntOrNull() ?: 0
    val pageable = PageRequest.of(pageNumber, pageSize, Sort.by(Sort.Direction.fromString(direction), sort))
    val result = enRayonService.getProduitsEnRayonPageableNew(
      nomProduit,
      bientotPerimeeBoolean,
      joursAvantPeremptionInt,
      enStockBoolean,
      pageable
    )
    return ResponseEntity.ok(result)
  }


  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/product/pageable/new")
  fun getProduitsEnRayonPageableProduitRange(
    @RequestParam(required = false) nomProduit: String?,
    @RequestParam(required = false) bientotPerimee: String?,
    @RequestParam(required = false) joursAvantPeremption: String?,
    @RequestParam(required = false) enStock: String?,
    @RequestParam(required = false) produitId: String?,
    @RequestParam(required = false) supprimer: String?,
    @RequestParam(required = false) startDate: String?,
    @RequestParam(required = false) endDate: String?,
    @RequestParam(defaultValue = "0") page: String,
    @RequestParam(defaultValue = "10") size: String,
    @RequestParam(defaultValue = "id") sort: String,
    @RequestParam(defaultValue = "desc") direction: String
  ): ResponseEntity<EnRayonPageableCustomDto> {
    val pageNumber = page.toIntOrNull()?.coerceAtLeast(0) ?: 0
    val pageSize = size.toIntOrNull()?.coerceAtLeast(1) ?: 10
    val bientotPerimeeBoolean: Boolean? = when (bientotPerimee?.lowercase()) {
      "true" -> true
      "false" -> false
      else -> null
    }
    val enStockBoolean: Boolean? = when (enStock?.lowercase()) {
      "true" -> true
      "false" -> false
      else -> null
    }
    val joursAvantPeremptionInt: Int = joursAvantPeremption?.toIntOrNull() ?: 0
    val pageable = PageRequest.of(pageNumber, pageSize, Sort.by(Sort.Direction.fromString(direction), sort))
    val result = enRayonService.getProduitsEnRayonPageableProduitRange(
      nomProduit,
      produitId,
      supprimer,
      startDate,
      endDate,
      bientotPerimeeBoolean,
      joursAvantPeremptionInt,
      enStockBoolean,
      pageable
    )
    return ResponseEntity.ok(result)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @DeleteMapping("/{id}")
  fun deleteEnRayon(@PathVariable id: Long): ResponseEntity<Void> {
    enRayonService.deleteEnRayon(id)
    return ResponseEntity.noContent().build()
  }

}

File: backend_py/pharmaPython/app/api/v1/en_rayon_controller.py
from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy.orm import Session
from typing import List, Dict, Any, Optional
from datetime import datetime


from app.api.deps import get_db
from app.models.en_rayon import EnRayon, EnRayonIn
from app.models.rayon import Rayon
from app.schemas.enrayon_dto import ProduitEnRayonDto, ProduitDetailIncrementEnRayonDto, EnRayonDto, \
  EnRayonPageableCustomDto
from app.services.en_rayon_service import EnRayonService
from app.utility.jwt_authentication import jwt_authentication

router = APIRouter(
  prefix="/en-rayon",
  tags=["EnRayon"],
  dependencies=[Depends(jwt_authentication)]  # équivalent @PreAuthorize
)

# ---------------------------
# Ajouter produits en rayon
# ---------------------------
@router.post("/ajouter", response_model=List[EnRayonIn])
def ajouter_produits(produits: List[ProduitEnRayonDto], db: Session = Depends(get_db)):
  service = EnRayonService(db)
  return service.ajouter_produits_en_rayon(produits)

# ---------------------------
# Mettre à jour produits en rayon
# ---------------------------
@router.put("/mettre-a-jour", response_model=List[EnRayonIn])
def mettre_a_jour_produits(produits: List[ProduitEnRayonDto], db: Session = Depends(get_db)):
  service = EnRayonService(db)
  return service.mettre_a_jour_produits_en_rayon(produits)

# ---------------------------
# Décrémenter stock
# ---------------------------
@router.post("/increment-produit-detail")
def decrementer_stock(data: ProduitDetailIncrementEnRayonDto, db: Session = Depends(get_db)):
  service = EnRayonService(db)
  return service.decrementer_stock(int(data.enRayonId), int(data.produitDetailId))

# ---------------------------
# Recherches par nom produit / produit / rayon / fournisseur
# ---------------------------
@router.get("/par-nom-produit", response_model=List[EnRayonIn])
def get_par_nom_produit(nomProduit: str, db: Session = Depends(get_db)):
  return EnRayonService(db).get_produits_par_nom_produit(nomProduit)

@router.get("/par-produit")
def get_par_produit(produitId: int, db: Session = Depends(get_db)):
  return EnRayonService(db).get_produits_par_produit(produitId)

@router.get("/par-produit-all")
def get_par_produit_all(produitId: int, produitType: str, db: Session = Depends(get_db)):
  return EnRayonService(db).get_produits_par_produit(produitId)

@router.get("/par-nom-rayon", response_model=List[EnRayonIn])
def get_par_nom_rayon(nomRayon: str, db: Session = Depends(get_db)):
  return EnRayonService(db).get_produits_par_nom_rayon(nomRayon)

@router.get("/par-fournisseur", response_model=List[EnRayonIn])
def get_par_fournisseur(nomFournisseur: str, db: Session = Depends(get_db)):
  return EnRayonService(db).get_produits_par_fournisseur(nomFournisseur)

@router.get("/par-commande", response_model=List[EnRayonIn])
def get_par_commande(commandeId: int, db: Session = Depends(get_db)):
  return EnRayonService(db).get_produits_par_commande(commandeId)

# ---------------------------
# Recherches par intervalle (dates & prix)
# ---------------------------
@router.get("/par-date-livraison", response_model=List[EnRayonIn])
def get_par_date_livraison(startDate: datetime, endDate: datetime, db: Session = Depends(get_db)):
  return EnRayonService(db).get_produits_par_intervalle_livraison(startDate, endDate)

@router.get("/par-date-peremption", response_model=List[EnRayonIn])
def get_par_date_peremption(startDate: datetime, endDate: datetime, db: Session = Depends(get_db)):
  return EnRayonService(db).get_produits_par_intervalle_peremption(startDate, endDate)

@router.get("/par-prix-achat", response_model=List[EnRayonIn])
def get_par_prix_achat(minPrix: float, maxPrix: float, db: Session = Depends(get_db)):
  return EnRayonService(db).get_produits_par_intervalle_prix_achat(minPrix, maxPrix)

@router.get("/par-prix-vente", response_model=List[EnRayonIn])
def get_par_prix_vente(minPrix: float, maxPrix: float, db: Session = Depends(get_db)):
  return EnRayonService(db).get_produits_par_intervalle_prix_vente(minPrix, maxPrix)

# ---------------------------
# Mise à jour simple (save)
# ---------------------------
@router.post("/save", response_model=EnRayonIn)
def mettre_a_jour_produit(rayonDto: EnRayonDto, db: Session = Depends(get_db)):
  return EnRayonService(db).mettre_a_jour_produit(rayonDto)

# ---------------------------
# Pagination
# ---------------------------
@router.get("/pageable")
def get_pageable(
  nomProduit: Optional[str] = None,
  bientotPerimee: Optional[str] = None,
  joursAvantPeremption: Optional[str] = None,
  enStock: Optional[str] = None,
  page: int = 0,
  size: int = 10,
  sort: str = "id",
  direction: str = "desc",
  db: Session = Depends(get_db),
):
  service = EnRayonService(db)
  return service.get_produits_pageable(
    nomProduit, bientotPerimee, joursAvantPeremption, enStock,
    page, size, sort, direction
  )

@router.get("/pageable/new")
def get_pageable_new(
  nomProduit: Optional[str] = None,
  bientotPerimee: Optional[str] = None,
  joursAvantPeremption: Optional[str] = None,
  enStock: Optional[str] = None,
  startDate: Optional[str] = None,
  endDate: Optional[str] = None,
  page: str = "0",
  size: str = "10",
  sort: str = "id",
  direction: str = "desc",
  db: Session = Depends(get_db),
):
  return EnRayonService(db).get_produits_pageable_new(
    nomProduit, bientotPerimee, joursAvantPeremption,
    startDate, endDate, enStock, page, size, sort, direction
  )

@router.get("/product/pageable/new")
def get_pageable_produit_range(
  nomProduit: Optional[str] = None,
  bientotPerimee: Optional[str] = None,
  joursAvantPeremption: Optional[str] = None,
  enStock: Optional[str] = None,
  produitId: Optional[str] = None,
  supprimer: Optional[str] = None,
  startDate: Optional[str] = None,
  endDate: Optional[str] = None,
  page: str = "0",
  size: str = "10",
  sort: str = "id",
  direction: str = "desc",
  db: Session = Depends(get_db),
):
  return EnRayonService(db).get_produits_pageable_produit_range(
    nomProduit, produitId, supprimer,
    startDate, endDate,
    bientotPerimee, joursAvantPeremption, enStock,
    page, size, sort, direction
  )

# ---------------------------
# Suppression
# ---------------------------
@router.delete("/{id}", status_code=204)
def delete_enrayon(id: int, db: Session = Depends(get_db)):
  EnRayonService(db).delete_enrayon(id)
  return {"message": "deleted"}

# ---------------------------
# Ajout automatique produits manquants
# ---------------------------
@router.get("/ajouter_tous_produits_manquant_en_rayon")
def ajouter_tous_manquants(db: Session = Depends(get_db)):
  return EnRayonService(db).ajouter_tous_produits_manquants()

@router.get("/ajouter_produit_manquant_en_rayon")
def ajouter_un_manquant(produitId: str, db: Session = Depends(get_db)):
  return EnRayonService(db).ajouter_un_produit_manquant(int(produitId))

@router.get("/reset_rayon")
def reset_negative_stock_to_zero(db: Session = Depends(get_db),
                                 produit_ids: str | None = None,
                                 reset_all: bool = False,
                                 only_negative:bool=False):
  return EnRayonService(db).reset_negative_stock_to_zero(
    produit_ids=[int(x.strip()) for x in produit_ids.split(",") if x.strip().isdigit()],
    reset_all=reset_all,
    only_negative=only_negative,
  )
